### PR TITLE
IR: Add sorting of data class equality checks depending on type

### PR DESCRIPTION
**Pull request by CommanderTvis**
Related issue is https://youtrack.jetbrains.com/issue/KT-23184

### CODE REVIEW

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/DataClassMembersGenerator.kt [hunk 1]
```diff
@@ -129,13 +130,21 @@ abstract class DataClassMembersGenerator(
...
-            for (property in properties) {
-                val arg1 = irGetProperty(irThis(), property)
-                val arg2 = irGetProperty(irGet(irType, otherWithCast.symbol), property)
-                +irIfThenReturnFalse(irNotEquals(arg1, arg2))
-            }
+
+            properties
+                .sortedBy {
+                    if (it.backingField!!.type.isPrimitiveType()) 1
+                    else 0
```
Comment (udalov) [inline compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/DataClassMembersGenerator.kt:140]: Wouldn't it result in non-primitive fields being compared first, and primitives second? Shouldn't it be the opposite?

**Review by udalov — CHANGES_REQUESTED** [review]
Please add a test with a data class with a bunch of fields of mixed types, with `equals` of non-primitive types having side effects (e.g. append some info to a global string), so that we could check precisely in what order the fields are compared.

**Comment by udalov** [comment]
As you mention yourself in KT-23184, it's a breaking change in behavior, so as with #4000, we can't push it right away, this needs to be approved by the Language Committee. So it has to be implemented under a `LanguageFeature`, tested both ways, etc.

**Comment by CommanderTvis** [comment]
@udalov How can I access supported features in that class?

**Comment by ilmirus** [comment]
@CommanderTvis You should pass an additional parameter to class constructors.

**Comment by dnpetrov** [comment]
Note that you can still do it uncoditionally if it's definitely known that equality comparisons for all data class property types are "pure" - e.g., `data class AllPure(val stringsArray: Array<String>, val int: Int)`.

**Comment by CommanderTvis** [comment]
I'm having some problems with building the branch (using `dist` task), so this PR is suspended.

image

**Pull request closed without merging**